### DISCUSSION TITLE
Is there any example of multi-modal MIL?

### DISCUSSION THREAD

**Discussion by filot997** [jamesdolezal/slideflow#341] [Q&A]
I've currently tried to write codes as follow.

config = mil_config(model="mm_attention_mil", batch_size=16, lr=1e-4)
P.train_mil(
        config=config,
        outcomes="category",
        train_dataset=train_dataset,
        val_dataset=val_dataset,
        bags="../models/v2/bags",
        input_header=columns.to_list()      
    )

But, I encounter error message such as "TypeError: _train_multimodal_mil() got an unexpected keyword argument 'input_header'"

So Is there any example that I want.

Thank you.

**Comment by jamesdolezal** [comment]
Hi filot977,

At present, the multimodal MIL is more accurately multi-_magnification_ MIL. With this approach, you can extract features at different magnifications (and/or different feature extractors). The `mm_attention_mil` architecture includes separate attention heads for each magnification / feature space. You can find an example of how to accomplish this [here](https://slideflow.dev/mil/#multi-magnification-mil). Briefly:

```python
# Configure a multimodal MIL model.
config = mil_config('mm_attention_mil', lr=1e-4)

# Set the bags paths for each modality.
bags_10x = '/path/to/bags_10x'
bags_40x = '/path/to/bags_40x'

P.train_mil(
    config=config,
    outcomes='HPV_status',
    train_dataset=train,
    val_dataset=val,
    bags=[bags_10x, bags_40x]
)
```

This architecture does *not*, however, support training with non-image features (e.g. clinical features). Extension of multi-modal support to allow for use of non-imaging features - like clinical variables, gene expression, and so on - is under development and anticipated to be released in the next 1-2 months.
